databay/misc/buffers.py

```diff
@@ -0,0 +1,75 @@
+import time
+from typing import List, Union
+
+from databay import Record
+
+
+class Buffer():
+    def __init__(self,
+                 count_threshold : int = None,
+                 time_threshold: float = None,
+                 custom_controllers: Union[callable, List[callable]] = None):
```

Comment:
Right now if one of the controllers returns true the buffer is released, the implementation is a boolean `OR` we could also have the option in the passed params for the buffer release that allows this to be set to a more strict `AND` so something like:

**For the __init__**
```python
    def __init__(self,
                 count_threshold: int = None,
                 time_threshold: float = None,
                 custom_controllers: Union[callable, List[callable]] = None,
                 strict_buffer_release: bool = False):
        self.count_threshold = count_threshold
        self.time_threshold = time_threshold
        self.custom_controllers = custom_controllers
        self.strict_buffer_release = strict_buffer_release
```

**For execute method**
```python
    def execute(self, records: List[Record]) -> List[Record]:
        self.records += records
        rv = []
        if self.flush:
            rv = self.records
        else:
            if self.strict_buffer_release:
                controllers_passing = True
                for controller in self.get_controllers():
                    if controller(self.records):
                        continue
                    else:
                        controllers_passing = False
                if controllers_passing:
                    rv = self.records

            else:
                for controller in self.get_controllers():
                    if controller(self.records):
                        rv = self.records

        if rv != []:
            self.reset()
        return rv

```

Reply:
Great suggestion! I briefly thought about that and wondered if you'd find that useful, so I'm glad to see that's the case 😊

Reply:
Great, so you put in a commit for this, will check it out now!